Firebase.cpp

```diff
@@ -76,180 +42,122 @@ Firebase& Firebase::auth(const String& auth) {
   return *this;
 }
 
-FirebaseGetResult Firebase::get(const String& path) {
-  FirebaseCall call(host_, auth_, "GET", path, &http_);
-  return call.isError() ? FirebaseGetResult::FromError(call.errorMessage())
-      : FirebaseGetResult::FromResponse(call.rawResponse());
+FirebaseGet Firebase::get(const String& path) {
+  return FirebaseGet(host_, auth_, path, &http_);
 }
 
-FirebasePushResult Firebase::push(const String& path, const String& value) {
-  FirebaseCall call(host_, auth_, "POST", path, value, &http_);
-  return call.isError() ? FirebasePushResult::FromError(call.errorMessage())
-      : FirebasePushResult::FromResponse(call.rawResponse());
+FirebasePush Firebase::push(const String& path, const String& value) {
+  return FirebasePush(host_, auth_, path, value, &http_);
 }
 
-FirebaseRemoveResult Firebase::remove(const String& path) {
-  FirebaseCall call(host_, auth_, "DELETE", path, &http_);
-  if (call.isError()) {
-    return FirebaseRemoveResult::FromError(call.errorMessage());
-  }
-  // Remove is only complete if returned status is OK(200).
-  if (!call.isOk()) {
-    return FirebaseRemoveResult::FromError(
-        "Remove " + path + " returned with status " + call.httpStatus());
-  }
-  return FirebaseRemoveResult::Ok();
+FirebaseRemove Firebase::remove(const String& path) {
+  return FirebaseRemove(host_, auth_, path, &http_);
 }
 
-FirebaseEventStream Firebase::stream(const String& path) {
-  return FirebaseEventStream(host_, auth_, path);
+FirebaseStream Firebase::stream(const String& path) {
+  return FirebaseStream(host_, auth_, path);  // stream doesn't reuse http client.
 }
 
-/* FirebaseCall */
+// FirebaseCall
 FirebaseCall::FirebaseCall(const String& host, const String& auth,
-                           const char* method, const String& path, const String& value,
-                           HTTPClient* http) : http_(http) {
-  const String url = makeUrl(path, auth);
-  http_->begin(host.c_str(), kFirebasePort, url.c_str(), true, kFirebaseFingerprint);
-  status_ =  http_->sendRequest(method, (uint8_t*)value.c_str(), value.length());
-  if (isError()) {
-    error_message_ = String(method) + " " + url + ": " + HTTPClient::errorToString(status_);
+			   const char* method, const String& path,
+			   const String& data, HTTPClient* http) {
+  if (!http) {
+    http = &http_;
```

Comment:
Always creating a HttpClient to only sometimes use seems wasteful but not a big deal I will fix this up down the road.